File: app/src/main/java/com/nan/xarch/eventbus/core/EventLiveData.java
package com.nan.xarch.eventbus.core;

import androidx.annotation.NonNull;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;

public class EventLiveData<T> extends MutableLiveData<T> {

    @Override
    public void observe(@NonNull LifecycleOwner owner, @NonNull Observer<? super T> observer) {
        this.observe(owner, false, observer);
    }

    @Override
    public void observeForever(@NonNull Observer<? super T> observer) {
        this.observeForever(false, observer);
    }

    public void observe(@NonNull LifecycleOwner owner, boolean sticky, @NonNull Observer<? super T> observer) {
        super.observe(owner, wrapObserver(sticky, observer));
    }

    public void observeForever(boolean sticky, @NonNull Observer<? super T> observer) {
        super.observeForever(wrapObserver(sticky, observer));
    }

    private Observer<T> wrapObserver(boolean sticky, Observer<? super T> observer) {
        return new EventObserverWrapper<>(this, sticky, observer);
    }

}

File: app/src/main/java/com/nan/xarch/eventbus/core/EventLiveData.kt
package com.nan.xarch.eventbus.core

import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer

class EventLiveData<T> : MutableLiveData<T>() {

    fun observe(owner: LifecycleOwner, sticky: Boolean, observer: Observer<in T>) {
        observe(owner, wrapObserver(sticky, observer))
    }

    fun observeForever(sticky: Boolean, observer: Observer<in T>) {
        observeForever(wrapObserver(sticky, observer))
    }

    private fun wrapObserver(sticky: Boolean, observer: Observer<in T>): Observer<T> {
        return EventObserverWrapper(this, sticky, observer)
    }
}